### PR TITLE
Update code generator to handle generic types and types with multiple '+' signs in them.

### PR DESCRIPTION
**Pull request by yaeldekel**
The C# code generator needs to be able to handle types that are nested inside generic classes. It should also handle types that are nested in nested classes.

### CODE REVIEW

**Comment by yaeldekel** [comment]
Fixes #101

File: src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs [hunk 1]
```diff
@@ -495,16 +495,56 @@ private void GenerateFooter(IndentingTextWriter writer)
             writer.WriteLine();
         }
 
-        static string GetSymbolFromType(Dictionary<string, string> typesSymbolTable, string fullTypeName, string currentNamespace)
+        static string GetSymbolFromType(Dictionary<string, string> typesSymbolTable, Type type, string currentNamespace)
```
Comment (TomFinley) [inline src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs]: >static [](start = 8, length = 6)

While we're at it let's add an explicit access modifier...

File: src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs [hunk 2]
```diff
@@ -495,16 +495,56 @@ private void GenerateFooter(IndentingTextWriter writer)
...
+
             if (fullTypeName.Contains('+'))
-                name += fullTypeName.Substring(0, fullTypeName.LastIndexOf('+')).Substring(fullTypeName.LastIndexOf('.') + 1);
+            {
+                int bracketIndex = fullTypeName.IndexOf('[');
+                if (bracketIndex > 0)
+                {
+                    Contracts.AssertValue(genericTypes);
+                    fullTypeName = fullTypeName.Substring(0, bracketIndex);
+                }
```
Comment (TomFinley) [inline src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs]: This is some fairly elaborate string manipulation going on, the utility of which will be opaque to anyone not intimately familiar with how the .NET class naming system works, and that also does not understand the specific issues raised in issue #101 that requires its introduction.

File: src/Microsoft.ML.Transforms/WhiteningTransform.cs [hunk 3]
```diff
@@ -571,7 +571,7 @@ private static Float DotProduct(Float[] a, int aOffset, Float[] b, int[] indices
 
         private static class Mkl
         {
-            private const string DllName = "Microsoft.ML.MklImports.dll";
+            private const string DllName = "Microsoft.MachineLearning.MklImports.dll";
```
Comment (TomFinley) [inline src/Microsoft.ML.Transforms/WhiteningTransform.cs]: >Microsoft.MachineLearning [](start = 44, length = 25)

Is this related to the issue raised?

Given that we decided on `Microsoft.ML` vs. the more verbose `Microsoft.MachineLearning`, should not that DLL change its name, rather than the other way around?

Comment (codemzs) [reply src/Microsoft.ML.Transforms/WhiteningTransform.cs]: I had the same question but I also feel this should be made as a separate change and we should open an issue for this.

---
In reply to: [187159434](https://github.com/dotnet/machinelearning/pull/102#discussion_r187159434) [](ancestors = 187159434)

File: src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs (hunk 2 above)
Comment (veikkoeeva) [reply src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs]: I have a feeling this is more complicated. For comparison, there's a similar case (maybe not exactly, but anyway): https://github.com/dotnet/orleans/blob/master/src/AdoNet/Orleans.Persistence.AdoNet/Storage/Provider/AdoNetGrainStorage.cs#L506. More laborius, but might be usable to extract this to a separate class/function and test a bit more. This is something that probably should be done in Orleans too, since there are similar needs. :)

Also perhaps interesting in case there's a possibility to refactor out commonalities and have a joint library: https://github.com/dotnet/orleans/blob/e78c7fa4ae302808ed30d71ebc14f1ef544a923e/src/Orleans.Core/AssemblyLoader/AssemblyLoader.cs.

Comment (yaeldekel) [reply src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs]: Hi Veikko, thanks for your suggestion and for the reference. This method definitely doesn't handle every possible case for a type name (for example, it does not handle the case where the nested class is also generic). Since it is a private method, I think that would be enough for handle the cases we need it to handle. I did add some comments though, to make it easier to handle new cases if the need arises.

File: src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs [hunk 4]
```diff
@@ -495,16 +495,58 @@ private void GenerateFooter(IndentingTextWriter writer)
...
         }
 
-        static string GetSymbolFromType(Dictionary<string, string> typesSymbolTable, string fullTypeName, string currentNamespace)
+        /// <summary>
+        /// This methods creates a unique name for a class/struct/enum, given a type and a namespace.
+        /// It generates the name based on the <see cref="Type.FullName"/> property of the type
+        /// (see description here https://msdn.microsoft.com/en-us/library/system.type.fullname(v=vs.110).aspx).
+        /// </summary>
+        /// <param name="typesSymbolTable">A dictionary containing the names of the classes already generated.
+        /// This parameter is only used to ensure that the newly generated name is unique.</param>
```
Comment (codemzs) [inline src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs]: It is not just used to ensure the newly generated name is unique but it contains the namespace to be used for a given type across the file. The goal of having a symbol table was to not have duplicate definition of the same types by nesting them in classes; this dictionary servers as a look up table whenever we want to reference a type.

File: src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs [hunk 5]
```diff
@@ -495,16 +495,58 @@ private void GenerateFooter(IndentingTextWriter writer)
...
+
+            // When the type is nested, the names of the outer types are concatenated with a '+'.
+            var nestedNames = fullTypeName.Split('+');
+            var baseName = nestedNames[0];
+            // We currently only handle generic types in the outer most class, support for generic inner classes
+            // can be added if needed.
+            int backTickIndex = baseName.LastIndexOf('`');
+            int dotIndex = baseName.LastIndexOf('.');
+            if (backTickIndex < 0)
+                name += baseName.Substring(dotIndex + 1);
```
Comment (codemzs) [inline src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs]: > name += baseName.Substring(dotIndex + 1); [](start = 15, length = 42)

Can dotindex be -1?

File: src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs [hunk 6]
```diff
@@ -495,16 +495,58 @@ private void GenerateFooter(IndentingTextWriter writer)
...
+                {
+                    foreach (var genType in genericTypes)
+                    {
+                        var splitNames = genType.FullName.Split('+');
+                        if (splitNames[0].LastIndexOf('.') >= 0)
+                            splitNames[0] = splitNames[0].Substring(splitNames[0].LastIndexOf('.') + 1);
+                        name += string.Join("", splitNames);
+                    }
+                }
+            }
```
Comment (codemzs) [inline src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs]: This is great, thanks for doing this. May be put an example of how a name is generated from this control flow.

File: src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs [hunk 7]
```diff
@@ -495,16 +495,58 @@ private void GenerateFooter(IndentingTextWriter writer)
...
 
-            name += fullTypeName.Substring(fullTypeName.LastIndexOf(dim) + 1); ;
+            for (int i = 1; i < nestedNames.Length; i++)
+                name += nestedNames[i];
```
Comment (codemzs) [inline src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs]: I'm assuming you are adding this to have unique names but I feel these names should go before the generated name and is it possible to only add them when we detect there is a name collision? We add them as needed and if we run out of them then we can add rolling numbers.

File: src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs (hunk 5 above)
Comment (TomFinley) [reply src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs]: If it somehow *cannot* be, we should add a nice little `Contracts.Assert` to document that..

---
In reply to: [187485225](https://github.com/dotnet/machinelearning/pull/102#discussion_r187485225) [](ancestors = 187485225)

File: src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs (hunk 2 above)
Comment (TomFinley) [reply src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs]: I might be tempted to insert a link though to the MSDN docs for this at the GetType, since we're attempting to read a (subset) of the language. Plus maybe document the limitations of this method.

---
In reply to: [187414843](https://github.com/dotnet/machinelearning/pull/102#discussion_r187414843) [](ancestors = 187414843)

File: src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs (hunk 5 above)
Comment (yaeldekel) [reply src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs]: I think not, because the full name always begins with the namespace.

---
In reply to: [187485225](https://github.com/dotnet/machinelearning/pull/102#discussion_r187485225) [](ancestors = 187485225)

File: src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs (hunk 7 above)
Comment (yaeldekel) [reply src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs]: Added an example in the XML comment.

---
In reply to: [187485666](https://github.com/dotnet/machinelearning/pull/102#discussion_r187485666) [](ancestors = 187485666)
